Add all qflags bases classes

bluebird75 opened the pull request:
I started from my work on windowFlags and look for ways to make it generic. I failed at defining a generic approach with mypy, protocols and generics inheriting from int, it is really above my understanding of mypy.

But, it's easy enough to generate a file for each QFlag based class which can be used to assert both the actual implementation types and values, and which ensures that the class is being correctly checked. This is what I am doing here for two classes : WindowFlags and Alignment.

With a few more hours, I should be able to cover all qflags based class. This would remove a big hole in PyQt stubs.

bluebird75 committed:
Fix WindowType and WindowFlags integer operations

Before the fix, the test would report:
windowflags.py:10: error: Incompatible types in assignment (expression has type "int", variable has type "WindowType")
windowflags.py:18: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:19: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:20: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:21: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:22: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:27: error: Unsupported operand types for + ("WindowFlags" and "WindowType")
windowflags.py:27: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:28: error: Unsupported operand types for - ("WindowFlags" and "WindowType")
windowflags.py:28: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:29: error: Unsupported operand types for | ("WindowFlags" and "WindowType")
windowflags.py:29: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:30: error: Unsupported operand types for & ("WindowFlags" and "WindowType")
windowflags.py:30: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:31: error: Unsupported operand types for ^ ("WindowFlags" and "WindowType")
windowflags.py:31: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:34: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:34: error: Unsupported operand types for + ("WindowType" and "WindowFlags")
windowflags.py:35: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:35: error: Unsupported operand types for - ("WindowType" and "WindowFlags")
windowflags.py:36: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:36: error: Unsupported operand types for | ("WindowType" and "WindowFlags")
windowflags.py:37: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:37: error: Unsupported operand types for & ("WindowType" and "WindowFlags")
windowflags.py:38: error: Incompatible types in assignment (expression has type "int", variable has type "WindowFlags")
windowflags.py:38: error: Unsupported operand types for ^ ("WindowType" and "WindowFlags")
windowflags.py:41: error: Unsupported left operand type for + ("WindowFlags")
windowflags.py:42: error: Unsupported left operand type for - ("WindowFlags")
windowflags.py:43: error: Unsupported left operand type for | ("WindowFlags")
windowflags.py:44: error: Unsupported left operand type for & ("WindowFlags")
windowflags.py:45: error: Unsupported left operand type for ^ ("WindowFlags")

bluebird75 committed:
Full testing on WindowType, WindowFlags and int combinations + correct annotaions

bluebird75 committed:
Update changelog

bluebird75 committed:
Make WindowType inherit int again (revert a previous change)

bluebird75 committed:
Ignore the errors reported from breaking Liskov principle

bluebird75 committed:
add one more test for windowFlags usage

bluebird75 committed:
Merge remote-tracking branch 'origin/master' into fix-windowflags-windowtype-operations

bluebird75 committed:
WindowType as working cross verfication with mypy and py.test

bluebird75 committed:
WindowFlags operations are verified by both mypy and pytest

bluebird75 committed:
Also test assignment operator

bluebird75 committed:
Fix CI: overload not needed, breaking Liskov is OK here

Breaking Liskov subsitution is OK because that's what PyQt5 does

bluebird75 committed:
Fix CI

bluebird75 committed:
Make stubtest happy: there is no need for a __init__ method on int subclass

bluebird75 committed:
Fix CI: functions are not needed

bluebird75 committed:
Make mypy happy

bluebird75 committed:
work in progress on making windowFlags test generic

bluebird75 committed:
Make test completely generic

bluebird75 committed:
Add tests and fixes for Alignment flags

bluebird75 committed:
Generator for testing QFlag based classes

bluebird75 committed:
concept of automating qflags validation

pfremy committed:
Sanitize input

pfremy committed:
Can associate a module with most QFlags

pfremy committed:
Look for qflags in all modules of qtbase

altendky commented on PyQt5-stubs/QtCore.pyi:
```diff
@@ -2755,13 +2761,19 @@ class Qt(sip.simplewrapper):
         def __invert__(self) -> 'Qt.Alignment': ...
         def __index__(self) -> int: ...
         def __int__(self) -> int: ...
+        def __or__ (self, other: typing.Union['Qt.Alignemtn', 'Qt.AlignmentFlag', int]) -> 'Qt.Alignemtn': ...
+        def __and__(self, other: typing.Union['Qt.Alignemtn', 'Qt.AlignmentFlag', int]) -> 'Qt.Alignemtn': ...
+        def __xor__(self, other: typing.Union['Qt.Alignemtn', 'Qt.AlignmentFlag', int]) -> 'Qt.Alignemtn': ...
+        def __ror__ (self, other: 'Qt.AlignmentFlag') -> 'Qt.Alignemtn': ...
+        def __rand__(self, other: 'Qt.AlignmentFlag') -> 'Qt.Alignemtn': ...
+        def __rxor__(self, other: 'Qt.AlignmentFlag') -> 'Qt.Alignemtn': ...
```
```suggestion
        def __or__ (self, other: typing.Union['Qt.Alignment', 'Qt.AlignmentFlag', int]) -> 'Qt.Alignment': ...
        def __and__(self, other: typing.Union['Qt.Alignment', 'Qt.AlignmentFlag', int]) -> 'Qt.Alignment': ...
        def __xor__(self, other: typing.Union['Qt.Alignment', 'Qt.AlignmentFlag', int]) -> 'Qt.Alignment': ...
        def __ror__ (self, other: 'Qt.AlignmentFlag') -> 'Qt.Alignment': ...
        def __rand__(self, other: 'Qt.AlignmentFlag') -> 'Qt.Alignment': ...
        def __rxor__(self, other: 'Qt.AlignmentFlag') -> 'Qt.Alignment': ...
```

bluebird75 replied:
I though I had fixed this one already. I think I was waiting on CI to see if it could pick it up

altendky committed:
Merge branch 'master' into all-qflags

altendky reviewed (dismissed):
I'll suggest this quickly now, in case you are around and we can get the CI fixed.  I'll go back to reviewing the test code now.

altendky commented on tests/generate_qflags_tests.py:
```diff
@@ -0,0 +1,69 @@
+FLAGS_FILES = [
+    ('qflags_alignmentflags.py', 'QtCore.Qt.Alignment', 'QtCore.Qt.AlignmentFlag', 'QtCore.Qt.AlignLeft', 'QtCore.Qt.AlignRight'),
+
+]
```
There would be other changes to go with this.  I prefer to use code to document things over comments when possible.

```suggestion
@dataclasses.dataclass
class FlagsFileNames:
    generated_file: str
    flags_class: str
    flag_class: str
    first_flag_value: str
    second_flag_value: str

FLAGS_FILES = [
    FlagsFileNames(
        generated_file='qflags_alignmentflags.py',
        flags_class='QtCore.Qt.Alignment',
        flag_class='QtCore.Qt.AlignmentFlag',
        first_flag_value='QtCore.Qt.AlignLeft',
        second_flag_value='QtCore.Qt.AlignRight',
    ),
]
```

bluebird75 replied:
This code is gone now, the super scripts replaces it.

altendky commented on tests/generate_qflags_tests.py:
```diff
@@ -0,0 +1,69 @@
+]
+
+# global values for simplicity of the code
+sourcePart1, sourcePart2, sourcePart3 = [], [], []  # type: Tuple[List[str], List[str], List[str]]
```
I'm a bit of a stickler for not using globals.  I also tend pretty strongly away from `if __name__ == "__main__":` checks, but I'll skip that one here.

1) It appears that `sourcePart2` is not output anywhere.  Am I just missing it?
2) Please no globals.
3) Let's have a CI check that the generated files are in fact up to date.  Maybe run the generation script and then `git diff --exit-code` against the generated files.  Or, just generate them before running tests and don't commit them.

bluebird75 replied:
this has been adressed now in the latest version

bluebird75 committed:
Update PyQt5-stubs/QtCore.pyi

Co-authored-by: Kyle Altendorf <sda@fstab.net>

bluebird75 commented:
The file where I would be interested for your review is qflags_aligmentflags.py . It uses a mostly generic approach on QFlag and QFlags based classes to check all operations on both mypy and python side.

Only two lines are specific to AlignementFlag in this file. I have work in progress to generate one similar file per QFlag based class (for the testing side) and applying automatically a fix to the pyi files to add the missing methods.

bluebird75 committed:
Remove unneeded #type: ignore and QtWidgets import to make CI happy

altendky committed:
Merge branch 'master' into all-qflags

altendky reviewed (changes requested):
I personally find the generated files to be pretty terrible.  But, to be clear, that is not a judgement of your work.  Rather, mypy is a static analyzer so it forces you to do things explicitly in ways that you wouldn't do them just for 'regular' Python code.  I've had a few ideas for alternative mechanisms to get mypy to check multiple types, but they don't seem like they will actually work.

I don't know what this should look like long term, other than upstream PyQt getting this all correct, but I can go with this for now.

bluebird75 commented:
Thanks, this is exactly the feedback I am looking for.

The use of globals was rather convenient when focusing on getting things done. I agree with you that it's better without it. Fully ok for using dataclasses, it certainly makes the thing more readable.

Just curious, how do you avoid the `if __name__ == '__main__'`  ? 

The sourcePart2 is indeed captured but not outputed anywhere. It is the part which is replaced.


For the full generation approach, I was thinking along the lines of : 
1. generate the qflag based test class
2. patch the relevant .pyi file
2. run local py.test + mypy test
3. if successful commit test file + updated pyi file, if fail, just skip the change as of now
4. move to next qflag class

One consequence of this generic approach is that there will be one test file par QFlag class, so expect quite some test files. Is that OK with you ? Or do you consider them ony as a local validation tool which does not need to go into git ?

bluebird75 committed:
work in progress

bluebird75 committed:
Work in progress for testing qflags

bluebird75 committed:
Can locate QFlag into pyqt module

bluebird75 committed:
improvements

bluebird75 committed:
Can parse pyi and check for enum class and its methods

bluebird75 committed:
work in progress: check for all method presence done

bluebird75 committed:
wip: module is modified for the enum class

bluebird75 committed:
wip: full code modification is working

bluebird75 committed:
wip: move code modification back to generate_qflags_stubs_and_tests

bluebird75 committed:
work in progress

bluebird75 committed:
Fix typo

bluebird75 committed:
Renaming and documentation update

bluebird75 committed:
work in progress

bluebird75 committed:
work in progress

bluebird75 committed:
Now working: running this script creates one more supported QFlag

bluebird75 committed:
Add all QFlag operations to QDBusConnection.ConnectionCabability in module QtDBus

bluebird75 committed:
Adding all QFlag operations to QGLFormat.OpenGLVersionFlags, QGLFormat.OpenGLVersionFlag in module QtOpenGL

bluebird75 committed:
Description of how to use the generation script

bluebird75 committed:
QFlag operations for QTextCodec.ConversionFlags, QTextCodec.ConversionFlag in module QtCore

bluebird75 committed:
More documentation and script cleaning

bluebird75 committed:
Restore the module when test fail

bluebird75 committed:
QFlag operations for QDir.Filters, QDir.Filter in module QtCore

bluebird75 committed:
QFlag operations for QDir.SortFlags, QDir.SortFlag in module QtCore

bluebird75 committed:
Merge branch 'all-qflags-phil' into all-qflags

bluebird75 commented:
The script is now ready. It was a bit more difficult than I expected but nothing too fancy. It is now 1000 lines, with a proper README. I will start committing some flags and see how it goes.

One issue which I am already encountering is that not all QFlags are equals. I encountered some small difference, which where caught by the strong test script. The script assumes one behavior and rejects everything else. This is fine for now, I'll look into the rejected cases another time to understand the behavior difference.

bluebird75 committed:
QFlag operations for QFileDevice.Permissions, QFileDevice.Permission in module QtCore

bluebird75 committed:
QFlag operations for QItemSelectionModel.SelectionFlags, QItemSelectionModel.SelectionFlag in module QtCore

bluebird75 committed:
QFlag operations for Qt.KeyboardModifiers, Qt.KeyboardModifier in module QtCore

bluebird75 committed:
QFlag operations for Qt.MouseButtons, Qt.MouseButton in module QtCore

bluebird75 committed:
QFlag operations for Qt.Orientations, Qt.Orientation in module QtCore

bluebird75 committed:
QFlag operations for Qt.WindowStates, Qt.WindowState in module QtCore

bluebird75 committed:
QFlag operations for Qt.ScreenOrientations, Qt.ScreenOrientation in module QtCore

bluebird75 committed:
QFlag operations for Qt.ImageConversionFlags, Qt.ImageConversionFlag in module QtCore

bluebird75 committed:
QFlag operations for Qt.DockWidgetAreas, Qt.DockWidgetArea in module QtCore

bluebird75 committed:
QFlag operations for Qt.ToolBarAreas, Qt.ToolBarArea in module QtCore

bluebird75 committed:
QFlag operations for Qt.Edges, Qt.Edge in module QtCore

bluebird75 committed:
Individual qflags test files no longer needed.

They are covered by the qflag automated tests

bluebird75 committed:
Less global variables

bluebird75 committed:
Re-run the failing commands to capture the actual error

bluebird75 committed:
QFlag operations for QGLContext.BindOptions, QGLContext.BindOption in module QtOpenGL

bluebird75 committed:
Add missing test file for WindowFlags / WindowType

bluebird75 committed:
QFlag test files for InputMethodQuery / InputMethodQueries

bluebird75 committed:
QFlag operations for Qt.InputMethodHints, Qt.InputMethodHint in module QtCore

bluebird75 committed:
QFlag operations for Qt.DropActions, Qt.DropAction in module QtCore

bluebird75 committed:
QFlag operations for Qt.ItemFlags, Qt.ItemFlag in module QtCore

bluebird75 committed:
QFlag operations for Qt.MatchFlags, Qt.MatchFlag in module QtCore

bluebird75 committed:
More explanation about --auto-commit

bluebird75 committed:
Merge branch 'all-qflags-phil' into all-qflags

bluebird75 committed:
QFlag operations for Qt.TextInteractionFlags, Qt.TextInteractionFlag in module QtCore

bluebird75 committed:
QFlag operations for Qt.TouchPointStates, Qt.TouchPointState in module QtCore

bluebird75 committed:
QFlag operations for Qt.GestureFlags, Qt.GestureFlag in module QtCore

bluebird75 committed:
QFlag operations for Qt.MouseEventFlags, Qt.MouseEventFlag in module QtCore

bluebird75 committed:
QFlag operations for QSurfaceFormat.FormatOptions, QSurfaceFormat.FormatOption in module QtGui

bluebird75 committed:
QFlag operations for QGL.FormatOptions, QGL.FormatOption in module QtOpenGL

bluebird75 committed:
Proper handling of flags in multiple modules or multiple times in the same module

bluebird75 committed:
QFlag operations for QOpenGLShader.ShaderType, QOpenGLShader.ShaderTypeBit in module QtGui

bluebird75 committed:
Merge branch 'all-qflags-phil' into all-qflags

bluebird75 committed:
QFlag operations for QMessageBox.StandardButtons, QMessageBox.StandardButton in module QtWidgets

bluebird75 committed:
QFlag operations for QDirIterator.IteratorFlags, QDirIterator.IteratorFlag in module QtCore

bluebird75 committed:
QFlag operations for QTreeWidgetItemIterator.IteratorFlags, QTreeWidgetItemIterator.IteratorFlag in module QtWidgets

bluebird75 committed:
QFlag operations for Qt.Alignment, Qt.AlignmentFlag in module QtCore

bluebird75 committed:
QFlag operations for Qt.InputMethodQueries, Qt.InputMethodQuery in module QtCore

bluebird75 committed:
QFlag operations for QEventLoop.ProcessEventsFlags, QEventLoop.ProcessEventsFlag in module QtCore

bluebird75 committed:
QFlag operations for QLibrary.LoadHints, QLibrary.LoadHint in module QtCore

bluebird75 committed:
QFlag operations for QOpenGLBuffer.RangeAccessFlags, QOpenGLBuffer.RangeAccessFlag in module QtGui

bluebird75 committed:
QFlag operations for QOpenGLDebugMessage.Sources, QOpenGLDebugMessage.Source in module QtGui

bluebird75 committed:
QFlag operations for QOpenGLDebugMessage.Severities, QOpenGLDebugMessage.Severity in module QtGui

bluebird75 committed:
QFlag operations for QMetaType.TypeFlags, QMetaType.TypeFlag in module QtCore

bluebird75 committed:
QFlag operations for QStandardPaths.LocateOptions, QStandardPaths.LocateOption in module QtCore

bluebird75 committed:
QFlag operations for QCommandLineOption.Flags, QCommandLineOption.Flag in module QtCore

bluebird75 committed:
QFlag operations for QTextOption.Flags, QTextOption.Flag in module QtGui

bluebird75 committed:
Store error on multiple lines for more readability

bluebird75 committed:
Fix QIODevice.OpenMode / OpenModeFlags

bluebird75 committed:
QFlag operations for QHstsPolicy.PolicyFlags, QHstsPolicy.PolicyFlag in module QtNetwork

bluebird75 committed:
Prepare for variations of OR support between QFlags

bluebird75 committed:
QFlag operations for QColorDialog.ColorDialogOptions, QColorDialog.ColorDialogOption in module QtWidgets

bluebird75 committed:
QFlag operations for QFontDialog.FontDialogOptions, QFontDialog.FontDialogOption in module QtWidgets

bluebird75 committed:
QFlag operations for QUrl.ComponentFormattingOptions, QUrl.ComponentFormattingOption in module QtCore

bluebird75 commented:
After some review, archiving a test file for each added QFlag is too much. There will be close to 100 test files just for the 10 modules listed today. When adding all other QFlags from all other modules, we should be close to 300 extra test files. This will just slow down the CI without any significant benefit. Maybe a lighter form to verify that the flag is properly supported should be the target, like checking one or two operations per flag.

bluebird75 committed:
QFlag operations for QWidget.RenderFlags, QWidget.RenderFlag in module QtWidgets

bluebird75 committed:
QFlag operations for QTextItem.RenderFlags, QTextItem.RenderFlag in module QtGui

bluebird75 committed:
QFlag operations for QPaintEngine.PaintEngineFeatures, QPaintEngine.PaintEngineFeature in module QtGui

bluebird75 committed:
QFlag operations for QTouchDevice.Capabilities, QTouchDevice.CapabilityFlag in module QtGui

bluebird75 committed:
QFlag operations for QOpenGLTexture.Features, QOpenGLTexture.Feature in module QtGui

bluebird75 committed:
QFlag operations for QPainter.RenderHints, QPainter.RenderHint in module QtGui

bluebird75 committed:
QFlag operations for QSizePolicy.ControlTypes, QSizePolicy.ControlType in module QtWidgets

bluebird75 committed:
QFlag operations for QRawFont.LayoutFlags, QRawFont.LayoutFlag in module QtGui

bluebird75 committed:
QFlag operations for QTextDocument.FindFlags, QTextDocument.FindFlag in module QtGui

bluebird75 committed:
QFlag operations for QHostAddress.ConversionMode, QHostAddress.ConversionModeFlag in module QtNetwork

bluebird75 committed:
QFlag operations for QAbstractSocket.BindMode, QAbstractSocket.BindFlag in module QtNetwork

bluebird75 committed:
QFlag operations for QAbstractSocket.PauseModes, QAbstractSocket.PauseMode in module QtNetwork

bluebird75 committed:
QFlag operations for QTextFormat.PageBreakFlags, QTextFormat.PageBreakFlag in module QtGui

bluebird75 committed:
QFlag operations for QNetworkInterface.InterfaceFlags, QNetworkInterface.InterfaceFlag in module QtNetwork

bluebird75 committed:
QFlag operations for QSsl.SslOptions, QSsl.SslOption in module QtNetwork

bluebird75 committed:
QFlag operations for QByteArray.Base64Options, QByteArray.Base64Option in module QtCore

bluebird75 committed:
QFlag operations for QLocale.NumberOptions, QLocale.NumberOption in module QtCore

bluebird75 committed:
QFlag operations for QDateTimeEdit.Sections, QDateTimeEdit.Section in module QtWidgets

bluebird75 committed:
QFlag operations for QRegularExpression.PatternOptions, QRegularExpression.PatternOption in module QtCore

bluebird75 committed:
QFlag operations for QRegularExpression.MatchOptions, QRegularExpression.MatchOption in module QtCore

bluebird75 committed:
QFlag operations for QTextStream.NumberFlags, QTextStream.NumberFlag in module QtCore

bluebird75 committed:
QFlag operations for QTextBoundaryFinder.BoundaryReasons, QTextBoundaryFinder.BoundaryReason in module QtCore

bluebird75 committed:
QFlag operations for QSql.ParamType, QSql.ParamTypeFlag in module QtSql

bluebird75 committed:
QFlag operations for QFileDialog.Options, QFileDialog.Option in module QtWidgets

bluebird75 committed:
QFlag operations for QFileIconProvider.Options, QFileIconProvider.Option in module QtWidgets

bluebird75 committed:
QFlag operations for QFileSystemModel.Options, QFileSystemModel.Option in module QtWidgets

bluebird75 committed:
QFlag operations for QGraphicsEffect.ChangeFlags, QGraphicsEffect.ChangeFlag in module QtWidgets

bluebird75 committed:
QFlag operations for QGraphicsBlurEffect.BlurHints, QGraphicsBlurEffect.BlurHint in module QtWidgets

bluebird75 committed:
QFlag operations for QInputDialog.InputDialogOptions, QInputDialog.InputDialogOption in module QtWidgets

bluebird75 committed:
QFlag operations for QWizard.WizardOptions, QWizard.WizardOption in module QtWidgets

bluebird75 committed:
QFlag operations for QAbstractItemView.EditTriggers, QAbstractItemView.EditTrigger in module QtWidgets

bluebird75 committed:
QFlag operations for QGraphicsItem.GraphicsItemFlags, QGraphicsItem.GraphicsItemFlag in module QtWidgets

bluebird75 committed:
QFlag operations for QGestureRecognizer.Result, QGestureRecognizer.ResultFlag in module QtWidgets

bluebird75 committed:
QFlag operations for QGraphicsScene.SceneLayers, QGraphicsScene.SceneLayer in module QtWidgets

altendky commented:
Do the tests actually take a long time?

bluebird75 committed:
Propagate the full class names to the final output

bluebird75 committed:
Merge branch 'all-qflags-phil' into all-qflags

bluebird75 committed:
QFlag operations for QGraphicsView.OptimizationFlags, QGraphicsView.OptimizationFlag in module QtWidgets

bluebird75 committed:
QFlag operations for QStyle.SubControls, QStyle.SubControl in module QtWidgets

bluebird75 committed:
QFlag operations for QAbstractSpinBox.StepEnabled, QAbstractSpinBox.StepEnabledFlag in module QtWidgets

bluebird75 committed:
QFlag operations for QStyleOptionFrame.FrameFeatures, QStyleOptionFrame.FrameFeature in module QtWidgets

bluebird75 committed:
QFlag operations for QStyleOptionButton.ButtonFeatures, QStyleOptionButton.ButtonFeature in module QtWidgets

bluebird75 committed:
QFlag operations for QStyleOptionTab.CornerWidgets, QStyleOptionTab.CornerWidget in module QtWidgets

bluebird75 committed:
QFlag operations for QStyleOptionToolBar.ToolBarFeatures, QStyleOptionToolBar.ToolBarFeature in module QtWidgets

bluebird75 committed:
QFlag operations for QStyleOptionViewItem.ViewItemFeatures, QStyleOptionViewItem.ViewItemFeature in module QtWidgets

bluebird75 committed:
QFlag operations for QStyleOptionToolButton.ToolButtonFeatures, QStyleOptionToolButton.ToolButtonFeature in module QtWidgets

bluebird75 committed:
QFlag operations for QDockWidget.DockWidgetFeatures, QDockWidget.DockWidgetFeature in module QtWidgets

bluebird75 committed:
QFlag operations for QFontComboBox.FontFilters, QFontComboBox.FontFilter in module QtWidgets

bluebird75 committed:
QFlag operations for QMdiArea.AreaOptions, QMdiArea.AreaOption in module QtWidgets

bluebird75 committed:
QFlag operations for QMdiSubWindow.SubWindowOptions, QMdiSubWindow.SubWindowOption in module QtWidgets

bluebird75 committed:
QFlag operations for QMainWindow.DockOptions, QMainWindow.DockOption in module QtWidgets

bluebird75 committed:
QFlag operations for QTextEdit.AutoFormatting, QTextEdit.AutoFormattingFlag in module QtWidgets

bluebird75 committed:
Display number of remaining flags to process

bluebird75 committed:
Also catch errors during transformation of the CST

bluebird75 committed:
Remove unneeded init method

bluebird75 committed:
Merge branch 'master' into all-qflags

bluebird75 committed:
Add mypy verifications of all qflags tests (and switch to Path API)

bluebird75 committed:
Accelerate mypy execution by grouping all qflags test files

bluebird75 committed:
Merge branch 'all-qflags' of https://github.com/bluebird75/PyQt5-stubs into all-qflags

bluebird75 committed:
Do not run mypy on test_stubs.py but on other files.

bluebird75 committed:
Check that qflag class inherits from sip.simplewrapper

bluebird75 committed:
Ensure filename part does not start with test_

bluebird75 committed:
QFlag operations for QStyle.State, QStyle.StateFlag in module QtWidgets

bluebird75 commented:
I was expecting a big impact but actually, after grouping all the qflag checks in one call, it is only 15 second impact. So very reasonable.

The branch is close to be ready for a merge. Anything else you want to see adjusted ?

bluebird75 committed:
Add ability to help finding the right class

bluebird75 committed:
Add changelog entry for qflags stubs

bluebird75 committed:
Adjust template to support more QFlag implementations

bluebird75 committed:
Support more QFlag behaviors

bluebird75 committed:
QFlag operations for QImageIOHandler.Transformations, QImageIOHandler.Transformation in module QtGui

bluebird75 committed:
QFlag operations for QTouchEvent.TouchPoint.InfoFlags, QTouchEvent.TouchPoint.InfoFlag in module QtGui

bluebird75 committed:
Generates test file and method according to QFlag behavior

bluebird75 committed:
QFlag operations for QNetworkProxy.Capabilities, QNetworkProxy.Capability in module QtNetwork

bluebird75 committed:
QFlag operations for QDBusConnection.RegisterOptions, QDBusConnection.RegisterOption in module QtDBus

bluebird75 committed:
QFlag operations for QDBusServiceWatcher.WatchMode, QDBusServiceWatcher.WatchModeFlag in module QtDBus

bluebird75 committed:
QFlag operations for QFileDevice.FileHandleFlags, QFileDevice.FileHandleFlag in module QtCore

bluebird75 committed:
QFlag operations for Qt.ApplicationStates, Qt.ApplicationState in module QtCore

bluebird75 committed:
QFlag operations for Qt.FindChildOptions, Qt.FindChildOption in module QtCore

bluebird75 committed:
QFlag operations for QUrl.UserInputResolutionOptions, QUrl.UserInputResolutionOption in module QtCore

bluebird75 committed:
QFlag operations for QNetworkConfiguration.StateFlags, QNetworkConfiguration.StateFlag in module QtNetwork

bluebird75 committed:
QFlag operations for QNetworkSession.UsagePolicies, QNetworkSession.UsagePolicy in module QtNetwork

bluebird75 committed:
QFlag operations for QPaintEngine.DirtyFlags, QPaintEngine.DirtyFlag in module QtGui

bluebird75 committed:
QFlag operations for QPainter.PixmapFragmentHints, QPainter.PixmapFragmentHint in module QtGui

bluebird75 committed:
QFlag operations for QGlyphRun.GlyphRunFlags, QGlyphRun.GlyphRunFlag in module QtGui

bluebird75 committed:
QFlag operations for QTextDocument.MarkdownFeatures, QTextDocument.MarkdownFeature in module QtGui

bluebird75 committed:
QFlag operations for QLocale.DataSizeFormats, QLocale.DataSizeFormat in module QtCore

bluebird75 committed:
QFlag operations for QAbstractPrintDialog.PrintDialogOptions, QAbstractPrintDialog.PrintDialogOption in module QtPrintSupport

bluebird75 committed:
QFlag operations for QPinchGesture.ChangeFlags, QPinchGesture.ChangeFlag in module QtWidgets

bluebird75 committed:
QFlag operations for QStyleOptionTab.TabFeatures, QStyleOptionTab.TabFeature in module QtWidgets

bluebird75 committed:
Fix dataclass incorrectly declared

bluebird75 committed:
wip: test file regeneration

bluebird75 committed:
QFlag operations for QDialogButtonBox.StandardButtons, QDialogButtonBox.StandardButton in module QtWidgets

bluebird75 committed:
Use a low value 1 for testing QFlag conversion from value

bluebird75 committed:
Result of processing the grep results. Basis for processing flags.

bluebird75 committed:
Regenerate grep results

bluebird75 committed:
Fix and improve grep results analysis

bluebird75 committed:
Better handling of missing expected attributes in classes

bluebird75 committed:
Improved grep results

bluebird75 committed:
Fix typo

bluebird75 committed:
QFlag operations for QOpenGLDebugMessage.Types, QOpenGLDebugMessage.Type in module QtGui

bluebird75 committed:
QFlag operations for QLocalServer.SocketOptions, QLocalServer.SocketOption in module QtNetwork

bluebird75 committed:
QFlag operations for QGraphicsView.CacheMode, QGraphicsView.CacheModeFlag in module QtWidgets

bluebird75 committed:
QFlag operations for QUrl.FormattingOptions, QUrl.UrlFormattingOption in module QtCore

Note that QUrl.FormattingOptions has a very limited implementation.

bluebird75 committed:
Add qflag support for CheckIndexOption in QtCore

bluebird75 committed:
README clarifications

bluebird75 committed:
Add result of our processing for history archiving

bluebird75 committed:
Message when we are finished

bluebird75 committed:
QFlag operations for QAbstractItemModel.CheckIndexOptions, QAbstractItemModel.CheckIndexOption in module QtCore

bluebird75 committed:
Notify when everything is finished.

bluebird75 committed:
Merge remote-tracking branch 'bluebird75/patch-2' into all-qflags-phil

bluebird75 committed:
Merge branch 'all-qflags-phil' into all-qflags

bluebird75 committed:
Add proper type: ignore[override] where needed

bluebird75 committed:
Remove __or__ and __ror__ for enum class

bluebird75 committed:
Remove __or__ and __ror__ where int base class already does the job.

bluebird75 committed:
Tell mypy to ignore unreachable code.

Unreachable code is here by design, because the template is multipurpose.

Also make the code 3.6 compatible by using typing_extensions

bluebird75 committed:
Int compatible enum do not need any changes.

bluebird75 committed:
Prevent mypy from reporting unreachable code (by design)

bluebird75 committed:
Regnerate test files to support Python 3.6 - 3.10

bluebird75 committed:
Regnerate test files to support Python 3.6 - 3.10

bluebird75 commented:
This is all green finally. Code is ready to be merged.

bluebird75 committed:
Merge branch 'master' into all-qflags

bluebird75 commented:
Hi Kyle. Can you start reviewing that one ? It is the result of a long work so I would be thrilled to have it integrated.

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
 def gen_tests():
-    for filename in os.listdir(TESTS_DIR):
-        if filename.endswith('.py') and not filename.startswith('test_'):
+    '''List of all tests files included in the directory tests'''
```
Docstrings specifically use double quotes.

```suggestion
    """List of all tests files included in the directory tests"""
```

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
             yield filename
 
+def gen_abs_qflags_tests():
+    '''List of all tests included in the directory qflags'''
```
```suggestion
    """List of all tests included in the directory qflags"""
```

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
+def gen_abs_qflags_tests():
+    '''List of all tests included in the directory qflags'''
+    for filename in (TESTS_DIR/'qflags').glob('test_*.py'):
+        yield filename
```
```suggestion
    yield from TESTS_DIR.joinpath('qflags').glob('test_*.py')
```

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
+    '''List of all tests files included in the directory tests'''
+    for filename in TESTS_DIR.glob('*.py'):
+        if not str(filename.parts[-1]).startswith('test_'):
             yield filename
```
`.parts` is already strings, but the last one is `.name` anyways.

```suggestion
    for path in TESTS_DIR.glob('*.py'):
        if not path.name.startswith('test_'):
            yield path
```

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
+
+def test_stubs_qflags() -> None:
+    """Run mypy over qflags files."""
+    stdout, stderr, exitcode = api.run([str(f) for f in gen_abs_qflags_tests()])
```
```suggestion
    stdout, stderr, exitcode = api.run([os.fspath(path) for path in gen_abs_qflags_tests()])
```

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
+
+@pytest.mark.parametrize('filename',
+                         list(gen_tests()),
+                         ids=[v.parts[-1] for v in gen_tests()]
```
Could be `.name` but this avoids prescribing a single directory being handled.  Also, `.as_posix()` will give a `str` with consistent forward slashes.
```suggestion
                         ids=[v.relative_to(TESTS_DIR).as_posix() for v in gen_tests()]
```

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
     """Run mypy over example files."""
-    path = os.path.join(TESTS_DIR, filename)
-    stdout, stderr, exitcode = api.run([path])
+    stdout, stderr, exitcode = api.run([str(filename)])
```
```suggestion
    stdout, stderr, exitcode = api.run([os.fspath(path)])
```

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
     assert not stderr
     assert exitcode == 0
 
+# note: no need to run explicitely pytest over qflags, because pytest finds them automatically
```
```suggestion
# note: no need to run explicitly pytest over qflags, because pytest finds them automatically
```

altendky commented on tests/test_stubs.py:
```diff
@@ -1,34 +1,60 @@
-    path = os.path.join(TESTS_DIR, filename)
-    with open(path, 'r') as f:
+    with open(str(filename), 'r') as f:
         code = f.read()
```
GitHub doesn't want me to suggest across the deleted line, but...

```python
    code = filename.read_text("r", encoding="utf-8")
```

altendky reviewed (approved):
Several remaining nitpicks about the use of `pathlib` (which I do very much like).  A general note I didn't point out everywhere is that I use `path` not `filename` for `pathlib.Path`s.  Feel free to merge after going through the suggestions.

For another PR, it would be good for the generation to be run in every CI run and for there to be a CI failure if the generation results in any changes, additions, or removals.

bluebird75 committed:
Update tests/test_stubs.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

bluebird75 committed:
Use triple quotes for docstrings

bluebird75 committed:
Update tests/test_stubs.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

bluebird75 committed:
Update tests/test_stubs.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

bluebird75 committed:
Update tests/test_stubs.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

bluebird75 committed:
Use os.fspath instead of str()

Co-authored-by: Kyle Altendorf <sda@fstab.net>

bluebird75 committed:
Typo

Co-authored-by: Kyle Altendorf <sda@fstab.net>

bluebird75 committed:
Improve style: use """ instead of ''', better use of Path, ...

bluebird75 committed:
Merge remote-tracking branch 'bluebird75/all-qflags' into all-qflags

bluebird75 commented:
Thanks for the feedback. This read_text() was unknown to me and I appreciated the other ideas.  The code was not written with top standards but to achieve a useful purpose. It probably could be improved.

It is not really feasible to run the generation on CI. First, it is very time consuming because libcst needs to parse the huge PyQt5 modules for each QFlag generation. Second, the source of the QFlag information is a manual grep over Qt sources, which requires some user modifications before being a proper source for complete generation.

But it's a good idea to run it on new Qt or PyQt releases.
